lib/std/zon/parse.zig

```diff
@@ -0,0 +1,2832 @@
+        for (0..nodes.len) |i| {
+            errdefer if (options.free_on_error) {
+                for (slice[0..i]) |item| {
+                    free(self.gpa, item);
+                }
+            };
+            slice[i] = try self.parseExpr(pointer.child, options, nodes.at(@intCast(i)));
+        }
+
+        return slice;
+    }
+
+    fn parseArray(
+        self: *@This(),
+        comptime T: type,
+        comptime options: Options,
+        node: Zoir.Node.Index,
+    ) !T {
+        const nodes: Zoir.Node.Index.Range = switch (node.get(self.zoir)) {
+            .array_literal => |nodes| nodes,
+            .empty_literal => .{ .start = node, .len = 0 },
+            else => return self.failExpectedContainer(T, node),
+        };
+
+        const array_info = @typeInfo(T).array;
+
+        // Check if the size matches
+        if (nodes.len > array_info.len) {
+            return self.failNodeFmt(
+                nodes.at(array_info.len),
+                "index {} outside of tuple length {}",
+                .{ array_info.len, array_info.len },
+            );
+        } else if (nodes.len < array_info.len) {
+            switch (nodes.len) {
+                inline 0...array_info.len => |n| {
+                    return self.failNodeFmt(node, "missing tuple field with index {}", .{n});
+                },
+                else => unreachable,
+            }
+        }
+
+        // Parse the elements and return the array
+        var result: T = undefined;
+        for (0..result.len) |i| {
+            // If we fail to parse this field, free all fields before it
+            errdefer if (options.free_on_error) {
+                for (result[0..i]) |item| {
+                    free(self.gpa, item);
+                }
+            };
+
+            result[i] = try self.parseExpr(array_info.child, options, nodes.at(@intCast(i)));
+        }
+        return result;
+    }
+
+    fn parseStruct(
+        self: *@This(),
+        comptime T: type,
+        comptime options: Options,
+        node: Zoir.Node.Index,
+    ) !T {
+        const repr = node.get(self.zoir);
+        const fields: @FieldType(Zoir.Node, "struct_literal") = switch (repr) {
+            .struct_literal => |nodes| nodes,
+            .empty_literal => .{ .names = &.{}, .vals = .{ .start = node, .len = 0 } },
+            else => return self.failExpectedContainer(T, node),
+        };
+
+        const field_infos = @typeInfo(T).@"struct".fields;
+
+        // Gather info on the fields
+        const field_indices = b: {
+            comptime var kvs_list: [field_infos.len]struct { []const u8, usize } = undefined;
+            inline for (field_infos, 0..) |field, i| {
+                kvs_list[i] = .{ field.name, i };
+            }
+            break :b std.StaticStringMap(usize).initComptime(kvs_list);
+        };
+
+        // Parse the struct
+        var result: T = undefined;
+        var field_found: [field_infos.len]bool = .{false} ** field_infos.len;
```

Comment:
```suggestion
        var field_found: [field_infos.len]bool = @splat(false);
```
Another one that came in after you started this branch; makes initializing big arrays much nicer.

Reply:
Oh nice, this is great